### ISSUE TITLE
Unsettled promises shouldn't be stored.

### ISSUE THREAD

**Issue by alcuadrado** [ethereumjs/ethereumjs-monorepo#1181]
When upgrading to v5 I discovered that there are multiple promises being stored, which is pretty dangerous and should be avoided. I've seen these two, but there's probably more:

* [`mclInitPromise` in the VM](https://github.com/ethereumjs/ethereumjs-monorepo/blob/486b992019572eb252ab93f150416a37f0d8162a/packages/vm/lib/index.ts#L24)
* [`initPromise` in the Blockchain](https://github.com/ethereumjs/ethereumjs-monorepo/blob/486b992019572eb252ab93f150416a37f0d8162a/packages/blockchain/src/index.ts#L127)

The reason this is dangerous is that they may be rejected before they get to be handled. In the case of `Blockchain`, there's [some code](https://github.com/ethereumjs/ethereumjs-monorepo/blob/486b992019572eb252ab93f150416a37f0d8162a/packages/blockchain/src/index.ts#L192-L194) that I believe tries to avoid it, but throwing in the `catch` of a promise just leads to a new one that's also rejected.

This is particularly problematic in Node 15+, as unhandled promise rejections now kill the process, even if the promise was stored. For example, this leads to a crash:

```js
class C {
  constructor() {
    this.promise = new Promise((resolve, reject) => {
      this.reject = reject;
    });
  }
}

const c = new C();
c.reject();

// This callback is meant to be executed in another event loop task
// which never gets to excecute. Node 15+ kills the process at the 
// end of a task that resulted in an unhandled promise rejection.
setImmediate(() => console.log("Here"));
```

How can these be done instead?

* In the case of `mclInitPromise`: I'd move its initialization into the opcode/precompile handlers that need them. Those are already async.
* In the case of `Blockchain`: Same as the VM, have an `init` method be called and await before every operation.

**Comment by holgerd77** [comment]
👍  I often had some uneasy feelings about this when seeing this code parts but couldn't put it into words what would have bothered me since I am not so firm with promises, thanks for this write-up. Would say that we should change here soon, we also have this pattern in some parts of the client.

**Comment by jochem-brouwer** [comment]
In both cases we could make the constructor of these classes `private` and instead provide a static constructor which awaits these initialization promises. (For VM we already have this, `VM.create`)

**Comment by alcuadrado** [comment]
Yeah, I really like that pattern. Even more than the `await this.init()` one.

**Comment by ryanio** [comment]
I am working on a PR to resolve this in `VM` and `Blockchain`.

Below is a working list of other places I've found with stored promises:

* devp2p:
  * [`Deferred`](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/devp2p/src/util.ts#L105-L115) class
* client:
  * [`VMExecution.vmPromise`](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/client/lib/sync/execution/vmexecution.ts#L14)
  * [`Fetcher.in, Fetcher.out`](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/client/lib/sync/fetcher/fetcher.ts#L52-L53) storing promises in `QHeap` (not sure if same concern applies here)
  * [`BoundProtocol.resolvers`](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/client/lib/net/protocol/boundprotocol.ts#L156-L157) storing promises

**Comment by acolytec3** [comment]
I'm digging through the synchronizer code this morning and saw this [`await this.blockFetcher...`](https://github.com/ethereumjs/ethereumjs-monorepo/blob/bfd53752d8d589a38fc3fb41bb67f5f5acb74f9d/packages/client/lib/sync/fullsync.ts#L133-L133) promise that looks like it just sits pending until sync completes (which could notionally be forever or close to it if we're syncing mainnet).  Does this count as a stored promise for purposes of this discussion?

**Comment by holgerd77** [comment]
Closed by #1811
